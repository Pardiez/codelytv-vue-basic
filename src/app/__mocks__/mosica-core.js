import { fakeGigsByDay } from './gigs-sample'
const mosicaCore = jest.genMockFromModule('mosica-core')

class GigService {
  retrieveNextGigs() {
    return Promise.resolve(fakeGigsByDay)
  }
}

mosicaCore.GigService = GigService
module.exports = mosicaCore
