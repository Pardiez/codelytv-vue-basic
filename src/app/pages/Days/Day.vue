<template>
  <q-card v-if="!isLoading">
    <div class='day'>
      <q-card-title>{{dateLabel}}</q-card-title>

      <div v-for="gig in day.gigs">
        <Gig :gig ="gig" :onClick="goTo"/>
      </div>
    </div>
  </q-card>
</template>

<script>
  import { MosicaRouter } from '../../services/MosicaRouter'
  import { spanishFromIso } from '../../services/date-utils'

  export default {
    props: {day: Object, isLoading: Boolean},
    created() {
      this.mosicaRouter = new MosicaRouter(this.$router)
    },
    methods: {
      goTo(gig) {
        this.mosicaRouter.navigateToGig(gig.id)
      }
    },
    computed: {
      dateLabel() {
        return spanishFromIso(this.day.date)
      }
    }
  }
</script>

<style scoped>
  .q-card-title {
    color: black;
    font-weight: bold;
    font-size: 150%;
  }
</style>
