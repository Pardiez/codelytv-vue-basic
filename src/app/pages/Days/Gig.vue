<template>
  <q-item @click="onClick(gig)" :key="gig.id" highlight link striped>
    {{gig.title}} - {{gig.place}}
  </q-item>
</template>

<script>
  export default {
    props: { gig: Object, onClick: Function }
  }
</script>
