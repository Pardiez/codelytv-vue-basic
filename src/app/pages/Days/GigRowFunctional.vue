<template functional>
  <q-item @click="props.onClick(props.gig)" :key="props.gig.id" highlight link striped>
    {{props.gig.title}} - {{props.gig.place}}
  </q-item>
</template>
