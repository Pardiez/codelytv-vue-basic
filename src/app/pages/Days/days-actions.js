import {
  FETCH_DAYS_REQUEST,
  FETCH_DAYS_SUCCESS,
  FETCH_DAYS_ERROR
} from './days-mutations'

import { retrieveDays } from '../../shared/mosica-api'

export function buildRetrieveDaysAction() {
  return retrieveDaysAction(retrieveDays).run
}

export function retrieveDaysAction(retrieveDays) {
  return { run }

  async function run({ commit }) {
    commit(FETCH_DAYS_REQUEST)
    try {
      const days = await retrieveDays()
      commit(FETCH_DAYS_SUCCESS, days)
    }
    catch (error) {
      commit(FETCH_DAYS_ERROR, error)
    }
  }
}

// Test to check how we can abstract common action patterns
export function retrieveDaysActionUsingQueryAction(retrieveDays) {
  return queryAction(retrieveDays, FETCH_DAYS_REQUEST, FETCH_DAYS_SUCCESS, FETCH_DAYS_ERROR)
}

export function queryAction(command, REQUEST, SUCCESS, ERROR) {
  return { run }

  async function run({ commit }) {
    commit(REQUEST)
    try {
      const result = await command()
      commit(SUCCESS, result)
    }
    catch (error) {
      commit(ERROR, error)
    }
  }
}
