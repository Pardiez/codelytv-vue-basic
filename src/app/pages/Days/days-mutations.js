export const FETCH_DAYS_REQUEST = 'fetchDaysRequest'
export const FETCH_DAYS_SUCCESS = 'fetchDaysSuccess'
export const FETCH_DAYS_ERROR = 'fetchDaysError'

export function fetchDaysRequest(state) {
  state.loading = true
  state.error = ''
}

export function fetchDaysSuccess(state, days) {
  state.loading = false
  state.days = days
  state.error = ''
}

export function fetchDaysError(state, error) {
  console.log('EMOS SIDO ENGAÃ‘ADO', error)
  state.loading = false
  state.error = error
}
