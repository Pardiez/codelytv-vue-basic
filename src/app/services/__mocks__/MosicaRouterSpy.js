import { MosicaRouter } from '../../services/MosicaRouter'
export let navigateToGigSpy = MosicaRouter().navigateToGig
navigateToGigSpy.mockClear()
