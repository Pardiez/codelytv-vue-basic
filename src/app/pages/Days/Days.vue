<template>
  <div class="full-width">
    <LoadSpinner :isLoading="isLoading"/>
    <div class="day" v-for="day in gigsByDay" v-if="!isLoading">
      <Day :day="day" :onClick="goTo"/>
    </div>
  </div>
</template>

<script>
  import { mapState } from 'vuex'
  import fetchJsonp from 'fetch-jsonp'
  import * as mosicaCore from 'mosica-core'
  import { HttpClient } from '../../services/HttpClient'
  import { MosicaRouter } from '../../services/MosicaRouter'
  const gigService = new mosicaCore.GigService(HttpClient(fetchJsonp), new mosicaCore.Matcher())

  export default {
    props: {
      gigService: {
        default: () => gigService
      }
    },
    data () {
      return {
        gigsByDay: [],
        isLoading: false,
        mosicaRouter: null
      }
    },
    async created() {
      this.isLoading = true
      this.gigsByDay = await this.gigService.retrieveNextGigs()
      this.isLoading = false
      this.mosicaRouter = new MosicaRouter(this.$router)
    },
    methods: {
      goTo(gig) {
        this.mosicaRouter.navigateToGig(gig.id)
      }
    },
    computed: {
      ...mapState(['gigs'])
    }
  }
</script>

<style>
</style>
