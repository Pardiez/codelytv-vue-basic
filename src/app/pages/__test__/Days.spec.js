import { mount } from 'vue-test-utils'
import Days from '@/app/pages/Days/Days.vue'
import { fakeGigsByDay } from '../../__mocks__/gigs-sample'
import { fakeGigsByDay2 } from '../../__mocks__/gigs-sample2'
import DayListPage from '../../__page_objects__/DaysPage'

jest.mock('../../services/MosicaRouter')
import { navigateToGigSpy } from '../../services/__mocks__/MosicaRouterSpy'

describe('Days', () => {
  const FIRST_DAY_GIG_TEXTS = fakeGigsByDay[0].gigs.map(
    gig => gig.title + ' - ' + gig.place)

  let page, wrapper
  beforeEach(async () => {
    wrapper = mount(Days)
    page = new DayListPage(wrapper)
    await page.updateAsync()
  })

  it('matches full snapshot', async() => {
    page.matchSnapshot()
  })

  it('renders all gigs in the first day', async() => {
    FIRST_DAY_GIG_TEXTS.map((text) => page.contains(text))
  })

  fit('navigates to first gig detail', async () => {
    const FIRST_GIG = fakeGigsByDay[0].gigs[0]
    page.clickFirstGig()
    expect(navigateToGigSpy).toHaveBeenCalledWith(FIRST_GIG.id)
  })

  fit('navigates to second gig detail', async () => {
    const SECOND_GIG = fakeGigsByDay[0].gigs[1]
    page.clickSecondGig()
    expect(navigateToGigSpy).toHaveBeenCalledWith(SECOND_GIG.id)
  })

  /* Different examples of more accurate tests that need
  to explicitly run over the DOM structure
  */

  it('render days', async () => {
    const DAYS = fakeGigsByDay.map((day) => day.day)
    expect(page.dayTexts()).toEqual(DAYS)
  })

  it('render gigs for each day', async () => {
    let expectedGigs
    fakeGigsByDay.map((day, index) => {
      expectedGigs = day.gigs.map((gig) => gig.title + ' - ' + gig.place)
      expect(page.gigRowsFor(index)).toEqual(expectedGigs)
    })
  })

  /* Test to demonstrate how to explicitly inject mosicaService as a prop
  (less MAGIC than manual jest mock)
   */
  it('render days(explicitly injected mock)', async () => {
    const gigService = {retrieveNextGigs: () => Promise.resolve(fakeGigsByDay2)}
    const wrapper = mount(Days, {propsData: {gigService}})
    page = new DayListPage(wrapper)
    await page.updateAsync()
    expect(wrapper.html()).toContain('rapsus')
  })

})
