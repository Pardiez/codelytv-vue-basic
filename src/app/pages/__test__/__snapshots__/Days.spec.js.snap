// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Days matches full snapshot 1`] = `
<div style="width: 100%;">
    <!---->
    <div class="day">
        <div class="q-list">
            <div class="day">
                <div class="q-list-header ">Lunes, 18 de Septiembre</div>
                <div>
                    <div class="q-item q-item-division relative-position q-item-highlight">
                        Anarchy in the JS - SCBCN
                    </div>
                </div>
                <div>
                    <div class="q-item q-item-division relative-position q-item-highlight">
                        Vegan Microservices - Around the world
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>
`;
